logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -0,0 +1,388 @@
+     * methods
+     */
+    private RubyHash convertArgs(final PluginDefinition def) {
+        final RubyHash converted = RubyHash.newHash(RubyUtil.RUBY);
+        for (final Map.Entry<String, Object> entry : def.getArguments().entrySet()) {
+            final Object value = entry.getValue();
+            final String key = entry.getKey();
+            final Object toput;
+            if (value instanceof PluginStatement) {
+                final PluginDefinition codec = ((PluginStatement) value).getPluginDefinition();
+                toput = pipeline.buildCodec(
+                    RubyUtil.RUBY.newString(codec.getName()),
+                    Rubyfier.deep(RubyUtil.RUBY, codec.getArguments())
+                );
+            } else {
+                toput = value;
+            }
+            converted.put(key, toput);
+        }
+        return converted;
+    }
+
+    /**
+     * Compiles a {@link RubyIntegration.Filter} from a given {@link PluginVertex}.
+     * @param vertex Filter {@link PluginVertex}
+     * @return Compiled {@link RubyIntegration.Filter}
+     */
+    private RubyIntegration.Filter buildFilter(final PluginVertex vertex) {
+        final PluginDefinition def = vertex.getPluginDefinition();
+        final SourceWithMetadata source = vertex.getSourceWithMetadata();
+        return pipeline.buildFilter(
+            RubyUtil.RUBY.newString(def.getName()), RubyUtil.RUBY.newFixnum(source.getLine()),
+            RubyUtil.RUBY.newFixnum(source.getColumn()), convertArgs(def)
+        );
+    }
+
+    /**
+     * Checks if a certain {@link Vertex} represents a {@link RubyIntegration.Filter}.
+     * @param vertex Vertex to check
+     * @return True iff {@link Vertex} represents a {@link RubyIntegration.Filter}
+     */
+    private boolean isFilter(final Vertex vertex) {
+        return filters.containsKey(vertex.getId());
+    }
+
+    /**
+     * Checks if a certain {@link Vertex} represents a {@link RubyIntegration.Output}.
+     * @param vertex Vertex to check
+     * @return True iff {@link Vertex} represents a {@link RubyIntegration.Output}
+     */
+    private boolean isOutput(final Vertex vertex) {
+        return outputs.containsKey(vertex.getId());
+    }
+
+    /**
+     * Compiles the next level of the execution from the given {@link Vertex} or simply return
+     * the given {@link Dataset} at the previous level if the starting {@link Vertex} cannot
+     * be expanded any further (i.e. doesn't have any more incoming vertices that are either
+     * a {code filter} or and {code if} statement).
+     * @param parents Nodes from the last already compiled level
+     * @param start Vertex to compile children for
+     * @return Datasets originating from given {@link Vertex}
+     */
+    private Collection<Dataset> flatten(final Collection<Dataset> parents, final Vertex start) {
+        final Collection<Vertex> endings = start.incomingVertices()
```

Comment:
This is a confusing variable name no? The `endings` are really the `startParents` right?

Reply:
hmm, agree, though I don't like `startParents` either :) I'll see if I can make this clearer some other way. No matter the variable name here, the method is kinda hard to understand.